### ISSUE TITLE
djangoForm service - no easy way to clear errors.

### ISSUE THREAD

**Issue by stuaxo** [jrief/django-angular#50]
djng_forms_module.factory('djangoForm', function() ...

Long version
In my app I've used the djangoForm service and pass it my own 'errors' which come back from the server after I have posted the data via ajax (which I do on each keypress, for a sort of ghetto instant apply).

This works, although there is no easy way to clear the data, so I cheat .. by passing setErrors an object containing all the field names and an empty array:

{ 
  field_name: [ "" ],
}

This is pretty dirty but works, the problem is that I get a tick after modifying a field, then when the server responds it disapears.

TLDR
Please can we have 'clearErrors'.

**Comment by fatzh** [comment]
I also have the same problem. I use djangoForm.setErrors() on ajax submit response, which works great and correctly adds the error to the fields, as well as the form.$error.rejected variable.

Now when I correct the error and submit again, if it's a success I use djangoForm.setErrors(form, null), but it doesn't work as the form.$error.rejected property is not there anymore. But the $message is still there for each field that previously failed. Because the main form.$error.rejected is no longer set, the loop to clean up previously set errors doesn't work.

Is there a proper way to clean the form after a successful ajax submission ? I tried to figure out when the form.$error.rejected variable gets cleaned up, but no luck..

**Comment by jrief** [comment]
Just to understand what's your requirement....
After a form has been submitted properly, and the Ajax handler returned signaling no error, then you want to set all form fields into an OK state, instead of loading another page.
Is this what you need?

**Comment by fatzh** [comment]
yes basically upon successful submission, I just display a success message next to the "save" button. It's a configuration form where it doesn't make sense to redirect the user to another page.

**Comment by jrief** [comment]
I just tried to reproduce your reports. What I found out is that the JS function ``setErrors()`` does its job properly, ie. it removes all exiting errors from a previously failed validation.
For this experiment I used http://djangular.aws.awesto.com/model_scope/
Could you please check this out locally and test with that demo. The only thing to do is to comment out this line: https://github.com/jrief/django-angular/blob/master/examples/server/static/js/model-scope.js#L7

**Comment by fatzh** [comment]
I'll try this asap and come back to you.

**Comment by fatzh** [comment]
ok I think I've found my problem.. I was using $setPristine in my success callback in order to detect new changes to the form and automatically discard the success message from the previous submission. It looks I had to use the setErrors() with empty error array before submitting the form and all works great now. Sorry about that.

**Comment by jrief** [comment]
@stuaxo 
did this help to solve your problem too? If so, please close.

**Comment by fatzh** [comment]
I've been looking a bit more into this and I know a bit more about the issue. I'm still investigating, as it's still shows up and I need (and expect) the form to be pristine after the successful ajax submission. The problem seems to be in the resetFieldValidity method, the `rejected` attribute gets cleared, but not the `$message`, making it to show up again when the form is pristine.

```javascript
function resetFieldValidity(field) {
	field.rejectedListenerPos = field.$viewChangeListeners.push(function() {
		// changing the field the server complained about, resets the form into valid state
		field.$setValidity('rejected', true);
		field.$viewChangeListeners.splice(field.rejectedListenerPos, 1);
		delete field.rejectedListenerPos;
	}) - 1;
}
```

If I add a `field.$message = "";` after the `field.$setValidity*"rejected", true);`, then it seems to work. 

```javascript
function resetFieldValidity(field) {
	field.rejectedListenerPos = field.$viewChangeListeners.push(function() {
		// changing the field the server complained about, resets the form into valid state
		field.$setValidity('rejected', true);
		// clearing error message from previous validation
		field.$message = '';
		field.$viewChangeListeners.splice(field.rejectedListenerPos, 1);
		delete field.rejectedListenerPos;
	}) - 1;
}
```

Am I doing something wrong ? I'm not sure why this problem doesn't happen on the demo.. The only difference that I can see is that I'm using a ModelForm.

**Comment by jrief** [comment]
I'm unable to reproduce this, but your description makes perfectly sense.
I'll fix that for the next version.
Thanks.

**Comment by fatzh** [comment]
I can provide a pull request if it helps.

**Comment by jrief** [comment]
don't bother for just one line of code

**Comment by fatzh** [comment]
ok cheers, and thanks again for the great module

**Comment by stuaxo** [comment]
Im not working on the project that uses django angular any more so cant test.. I used the empty array trick IIRC... close if the fix in next version is going in :)

**Comment by jrief** [comment]
@stuaxo OK, I'm closing this issue.
Just for curiosity, can you please point me on the URL which uses this library.

**Comment by stuaxo** [comment]
Would love to, unfortunately it was an intranet site, so will never be public.

The site was a backend system for administaring Nectar cards, the UI was pretty complex, but before that they had no UI, so seemed pretty happy with the result.
